### PR TITLE
feat: adding preference sheets to lottery export

### PR DESCRIPTION
**Pull request by YazeedLoonat**
This PR addresses #4056 

- [x] Addresses the issue in full

## Description
This adds the preference sheets to the lottery export as well as a few other things:

- preference sheets
    - added
- demographics
    - now based on role just like application csv export
- columns
    - added preference rank
    - still need to double check column order
- permissions
    - UI is now hooked up to do a partner export

## How Can This Be Tested/Reviewed?
- head to the partners site
- on a closed listing with applications and preferences on the listing (and to make it easier claimed preferences on at least some of the applications)
- generate the lottery results
- get the spreadsheet export

you should notice individual sheets for each preference as well as the original raw sheet

## Author Checklist:

- [ ] Added QA notes to the issue with applicable URLs
- [x] Reviewed in a desktop view
- [ ] Reviewed in a mobile view
- [ ] Reviewed considering accessibility
- [ ] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [ ] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

### CODE REVIEW

**Review by mcgarrye — DISMISSED** [review]
Raw rank on the preference sheet is the same for all applicants
Screenshot 2024-08-06 at 10 36 54 AM

**Comment by ludtkemorgan** [comment]
@mcgarrye the bug with all showing the same rank has been fixed. There was a `=` instead of `===`. I'll also add some tests around that

**Comment by emilyjablonski** [comment]
The [sample export](https://docs.google.com/spreadsheets/d/1dX9ehAq3pdMpmhNq1arWV2bwjHBFMy6xAbcPGhLVHOY/edit?gid=520192447#gid=520192447) has a slightly different column order with the raw rank coming first

File: api/src/services/lottery.service.ts [hunk 1]
```diff
@@ -426,22 +447,46 @@ export class LotteryService {
...
     queryParams: ApplicationCsvQueryParams,
     forLottery = false,
+    preference?: IdDTO,
   ): Promise<void> {
-    // create raw rank spreadsheet
-    const rawRankSpreadsheet = workbook.addWorksheet('Raw');
-    rawRankSpreadsheet.columns = this.buildExportColumns(csvHeaders);
+    // create a spreadsheet. If the preference is passed in use that as a title otherwise 'raw'
+    const spreadsheet = workbook.addWorksheet(
+      preference ? preference.name : 'Raw',
```
Comment (emilyjablonski) [inline api/src/services/lottery.service.ts:463]: The sample export has the title as `Raw Lottery Rank`

File: sites/partners/src/pages/listings/[id]/lottery.tsx [hunk 2]
```diff
@@ -44,7 +44,21 @@ const Lottery = (props: { listing: Listing }) => {
...
-  const { onExport, exportLoading } = useLotteryExport(listing?.id)
+
+  const listingJurisdiction = profile?.jurisdictions?.find(
+    (jurisdiction) => jurisdiction.id === listing?.jurisdictions.id
+  )
+
+  const includeDemographicsPartner =
+    profile?.userRoles?.isPartner && listingJurisdiction?.enablePartnerDemographics
+
+  const { onExport, exportLoading } = useLotteryExport(
```
Comment (emilyjablonski) [inline sites/partners/src/pages/listings/[id]/lottery.tsx:55]: Couldn't we do this all on the backend by looking at the requesting user?

Comment (ludtkemorgan) [reply sites/partners/src/pages/listings/[id]/lottery.tsx:55]: This is where we call the backend. Are you referring to the includeDemographics logic? The include demographics query param follows the same pattern as the csv export. We could move it to the backend but then we would also need to look up the jurisdiction to know if partnerdemographics is enabled

Comment (emilyjablonski) [reply sites/partners/src/pages/listings/[id]/lottery.tsx:55]: Yeah we don't have to make this change here, but I think it makes more sense to do all of that on the backend

File: api/src/services/lottery.service.ts (hunk 1 above)
Comment (ludtkemorgan) [reply api/src/services/lottery.service.ts]: fixed!

Comment (mcgarrye) [reply api/src/services/lottery.service.ts]: It seems like you changed the wrong header name :/
Screenshot 2024-08-12 at 11 43 23 AM

**Comment by ludtkemorgan** [comment]
@mcgarrye nice catch. I have fixed the columns to have the right data

**Pull request merged**